src/terms/Hamiltonian.jl

```diff
@@ -26,13 +26,18 @@ struct Hamiltonian
     blocks::Vector{HamiltonianBlock}
 end
 
+
 # Loop through bands, IFFT to get ψ in real space, loop through terms, FFT and accumulate into Hψ
-# Do this band by band to conserve memory
-# As an optimization we special-case nonlocal operators to apply them
-# instead on the full block and benefit from BLAS3
 @views @timing "Hamiltonian multiplication" function LinearAlgebra.mul!(Hψ::AbstractArray,
                                                                         H::HamiltonianBlock,
                                                                         ψ::AbstractArray)
+    if length(H.optimized_operators) == 3 &&
+        any(o -> o isa FourierMultiplication, H.optimized_operators) &&
+        any(o -> o isa RealSpaceMultiplication, H.optimized_operators) &&
+        any(o -> o isa NonlocalOperator, H.optimized_operators)
+        return mul!_fast(Hψ, H, ψ)
+    end
```

Comment:
This is not very Julian. What I would do is have a special type, e.g.
```
Base.@kwdef struct DftOperators
fourier_multiplication
real_space_multiplication
nonlocal_operator
end
```
and then pass `mul!` over to a four-argument version that also has the `H.optimized_operators` as an argument. Than you can dispatch on that to implement the special `mul!` function. The check for that case (and the introduction of that special type) can be done in `optimise_operators_`

Reply:
That just adds more indirection, more code and more complexity. I like the current hack better: it's clear what's going on here: there's a generic path, and a fast optimized path.

Reply:
The part I don't like is that you manually do multiple dispatch here. Also these 6 lines of check code are not exactly fun to read. Also `mul!_fast` is much less meaningful than having `mul!( ... , ::DftOperators, ...)` somewhere. And code complexity? The type can be defined pretty locally in the same file. I think when it comes to number of code lines its no longer that what you have now it's probably even shorter and you don't have to do the weird extraction stuff like `https://github.com/JuliaMolSim/DFTK.jl/pull/354/files/5cfbb2a4f3261093c23a7b1978fa9e5da0d75de0#diff-c88cddd6ac7d48004a07a50ffcaf0f3877f04610fd6858f13a23cccae8035d06R77` again, which is also complex code.

Reply:
It's a hack. We don't want to institutionalize hacks by promoting them to have data structures, the simplest they are the better. I thought about having an extra `eligible_for_fast_path` or so function, but it's not really better...

Reply:
Ok, I agree with that. My main point is that I disapprove of these 6 lines of explicit check code and that weird extraction routine and the `mul!_fast` as a name. What I propose solves all of these. I'm ok to not introduce the type if we come up with a more elegant solution to solve these aspects.

Re `eligible_for_fast_path` ... at some point you will have to check and I think a good point is when you construct the Hamiltonian and anyway get your hands dirty with the operators by calling `optimise_operators` don't you think? So why not determine that info there and somehow pass it on?

Reply:
I pushed something. I'm not sure I like it too much, it's more code, but if you prefer we can keep it.